File: backend/src/main/java/com/back/ourlog/global/security/exception/CustomAccessDeniedHandler.java
package com.back.ourlog.global.security.exception;

import com.back.ourlog.global.common.dto.RsData;
import com.back.ourlog.global.exception.ErrorCode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

@Component
@RequiredArgsConstructor
public class CustomAccessDeniedHandler implements AccessDeniedHandler {

    private final ObjectMapper objectMapper;

    @Override
    public void handle(HttpServletRequest request, HttpServletResponse response,
                       AccessDeniedException accessDeniedException) throws IOException {

        response.setContentType("application/json;charset=UTF-8");
        response.setStatus(HttpServletResponse.SC_FORBIDDEN);

        RsData<Void> body = RsData.fail(ErrorCode.AUTH_FORBIDDEN);

        String json = objectMapper.writeValueAsString(body);
        response.getOutputStream().write(json.getBytes(StandardCharsets.UTF_8));
    }
}

File: backend/src/main/java/com/back/ourlog/global/security/exception/CustomAccessDeniedHandler.kt
package com.back.ourlog.global.security.exception

import com.back.ourlog.global.common.dto.RsData.Companion.fail
import com.back.ourlog.global.exception.ErrorCode
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets

@Component
class CustomAccessDeniedHandler (
    private val objectMapper: ObjectMapper
) : AccessDeniedHandler {

    override fun handle(
        request: HttpServletRequest?, response: HttpServletResponse,
        accessDeniedException: AccessDeniedException
    ) {
        response.contentType ="application/json;charset=UTF-8"
        response.status = HttpServletResponse.SC_FORBIDDEN

        val body = fail<Void?>(ErrorCode.AUTH_FORBIDDEN)

        val json = objectMapper.writeValueAsString(body)
        response.outputStream.write(json.toByteArray(StandardCharsets.UTF_8))
    }
}
